Add VK_EXT_nested_command_buffer support

james-sumihiro opened the pull request:
Added support for VK_EXT_nested_command_buffer

baldurk commented on renderdoc/driver/vulkan/extension_support.md:
```diff
@@ -75,6 +75,7 @@ Maintainers can update this file by updating vk.xml in this folder and running `
 * `VK_EXT_metal_surface`
 * `VK_EXT_multisampled_render_to_single_sampled`
 * `VK_EXT_mutable_descriptor_type`
+* `VK_EXT_nested_command_buffer`
```
This should also remove it from the 'unsupported' lists below (~line 286).

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -4173,7 +4176,7 @@ void WrappedVulkan::ReplayLog(uint32_t startEventID, uint32_t endEventID, Replay
     VkResult vkr = VK_SUCCESS;
 
-    bool rpWasActive[2] = {};
+    std::map<ResourceId, Submission> prevSubmissionState = {};
```
Could you please use `rdcflatmap` here rather than `std::map`? It's not a 100% drop-in replacement but should work for the cases you have here. It's tailored for cases like this where we expect that most of the time there will only be one or maybe two elements so it just boils down to a plain `rdcarray` with linear lookup.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -4202,10 +4205,20 @@ void WrappedVulkan::ReplayLog(uint32_t startEventID, uint32_t endEventID, Replay
+
+      bool rpActive = false;
+
+      for(const auto &partialSubmit : prevSubmissionState)
```
Please prefer using an explicit type here (`const rdcpair<ResourceId, Submission> &` if you change to rdcflatmap) rather than `auto`.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -4285,14 +4298,20 @@ void WrappedVulkan::ReplayLog(uint32_t startEventID, uint32_t endEventID, Replay
+      {
+        m_Partial.partialSubmits[partialSubmit.first].renderPassActive =
+            partialSubmit.second.renderPassActive;
+      }
```
You've lost the comment here explaining why the `renderPassActive` state is restored.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -4285,14 +4298,20 @@ void WrappedVulkan::ReplayLog(uint32_t startEventID, uint32_t endEventID, Replay
+          m_RenderState.EndRenderPass(cmd);
+          break;
+        }
+      }
```
I think if `rpActive` is kept at the same scope as `rpWasActive` is, we can re-use it here to check the condition without looping here?

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -4779,16 +4798,9 @@ bool WrappedVulkan::InRerecordRange(ResourceId cmdid)
-             m_LastEventID;
-    }
-  }
+  if(m_Partial.partialSubmits.count(cmdid))
```
I wasn't aware of `std::map::count()` before, that's pretty handy for a "is this key present" check without needing to check an iterator. However in this case we will want the value below, so you should do the usual thing of `find` to return an iterator to avoid two lookups.

This has also deleted the comment, which to me still seems like it applies since it doesn't specifically refer to primary/secondary but just "one of the [..] partial command buffers".

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -1171,48 +1171,37 @@ bool WrappedVulkan::Serialise_vkBeginCommandBuffer(SerialiserType &ser, VkComman
-#endif
+          m_Partial.currentPartial = BakedCommandBuffer;
+          m_Partial.isPartialPrimary =
+              (m_BakedCmdBufferInfo[BakedCommandBuffer].level == VK_COMMAND_BUFFER_LEVEL_PRIMARY);
```
This is not going to work in all cases depending on which partial command buffer we want to track. As you've mentioned there are up to N different command buffers (with a nesting depth of N) that are partial at any given event, but they can be begun in any order as long as secondaries are ended before being executed in a parent. Overwriting the current here will give you effectively a random ordering depending on the order the application decided to call begin.

Note this interacts with other feedback I've put on the definition of the `PartialReplayData`.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -1171,48 +1171,37 @@ bool WrappedVulkan::Serialise_vkBeginCommandBuffer(SerialiserType &ser, VkComman
 
-      // check for partial execution of this command buffer
-      for(int p = 0; p < ePartialNum; p++)
+      for(const auto &submitEvent : submitEvents)
```
I won't comment on every case like this, but just to clarify please prefer a concrete type rather than `auto` except in the case of stl iterators like `std::map<blah, blah>::iterator`. In this case since the inner element is just `uint32_t` I believe it's better to have it iterated by value than const reference.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -1425,10 +1414,12 @@ VkResult WrappedVulkan::vkBeginCommandBuffer(VkCommandBuffer commandBuffer,
-                                          eFrameRef_Read);
-      record->MarkResourceFrameReferenced(GetResID(pBeginInfo->pInheritanceInfo->framebuffer),
-                                          eFrameRef_Read);
+      if(pBeginInfo->pInheritanceInfo->renderPass != VK_NULL_HANDLE)
```
Is this check necessary? `GetResID(VK_NULL_HANDLE)` returns a plain `ResourceId()` and `MarkResourceFrameReferenced` should automatically drop that. It saves a call but this should be inlined anyway.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -4875,12 +4880,14 @@ bool WrappedVulkan::Serialise_vkCmdExecuteCommands(SerialiserType &ser, VkComman
-        uint32_t startEID = parentCmdBufInfo.curEventID + m_Partial[Primary].baseEvent;
+        uint32_t startEID = parentCmdBufInfo.curEventID;
+
+        if(m_Partial.partialSubmits.count(m_LastCmdBufferID) != 0)
```
Again I think a plain `find` here is better to re-use the iterator and avoid a second lookup below.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -7396,13 +7392,9 @@ bool WrappedVulkan::Serialise_vkCmdEndRendering(SerialiserType &ser, VkCommandBu
           m_BakedCmdBufferInfo[m_LastCmdBufferID].renderPassOpen = false;
 
-          // if this rendering is just being suspended, the pass is still active
           if(!suspending)
```
Again there's a deleted comment here and I don't know why - it still seems entirely valid?

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp:
```diff
@@ -906,6 +892,50 @@ void WrappedVulkan::InsertActionsAndRefreshIDs(BakedCmdBufferInfo &cmdBufInfo)
+  {
+    cmdsWithReferences.push_back(subcmd->bakedCommands);
+
+    subcmd->bakedCommands->AddReferencedIDs(refdIDs);
```
I would like to preserve the comment in some way that explains why we're not including frame references here and instead pushing into `cmdsWithReferences`. This bit of the code is pretty hairy so any clarity is helpful.

baldurk replied:
As in this note I'd still like to see a comment locally inside this function explainining the `cmdsWithReferences`. It duplicates the comment at the call site but there's no cost to that and it makes it more likely someone will understand the code in future.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -4875,12 +4880,14 @@ bool WrappedVulkan::Serialise_vkCmdExecuteCommands(SerialiserType &ser, VkComman
-        // command buffer
-        if(m_Partial[Primary].partialParent != m_LastCmdBufferID)
+        // command buffers
+        if(m_Partial.isPartialPrimary && m_Partial.currentPartial != m_LastCmdBufferID)
```
I'm not sure this check wants to be restricted to only executes primary command buffers.

In the re-record range if we see an execute in a secondary which is not in the partial set, we also want to fully re-record it.

baldurk commented on renderdoc/driver/vulkan/vk_core.h:
```diff
@@ -735,59 +746,44 @@ class WrappedVulkan : public IFrameCapturer
+    // where we need to save the state of the partial stack and restore it (see ReplayLog())
+    std::map<ResourceId, Submission> partialSubmits;
+    bool isPartialPrimary;
+    ResourceId currentPartial;
```
I don't think that storing only one command buffer here is sufficient.

This information is used for only a couple of things in the end:

1. For knowing if this is the (single) primary partial command buffer. For updating RP state via `WrappedVulkan::ShouldUpdateRenderState()` and for closing any pending partially-replayed things like render passes, XFB, conditional rendering.
2. For knowing the _innermost_ partial command buffer (whether it be primary or secondary) to know which one to pull render state from via `WrappedVulkan::GetPartialCommandBuffer()`.
3. For knowing if this is _any_ partial command buffer to close a pending dynamic rendering instance which could have been begun in any secondary command buffer.

So I think that means we need to have an equivalent knowledge to previously of knowing the full list of N partial command buffers all the way down to the innermost.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -4811,37 +4823,27 @@ bool WrappedVulkan::ShouldUpdateRenderState(ResourceId cmdid, bool forcePrimary)
-    return cmdid == m_Partial[Secondary].partialParent;
-
-  return cmdid == m_Partial[Primary].partialParent;
+  return cmdid == m_Partial.currentPartial;
```
As an aside, I went digging on this function because it seemed very wrong only to realise that it doesn't really do what it says anymore, and these comments are inaccurate.

From history diving the relevant commits are 41d1acc47a040bd009d9d24f105a07d99413d4b2 and ef2804bfd30898816409f9a6541a1448969cafb5. Previously we used this function to do as it says - check if the renderstate should be updated (only for the innermost partial command buffer unless forced to the primary, as this does), but then was refactored in favour of per-commandbuffer renderstate tracking which could be done unconditionally.

I would suggest this function should be renamed to `IsPartialCommandBuffer` and perhaps even split out the `forcePrimary` to a separate function like `IsPrimaryPartialCommandBuffer`, since it's only now used for tracking renderpass activity, and not state, so the name is quite misleading.

baldurk reviewed (changes requested):
I think this is the right direction but I'm not sure the current partial parent tracking is sufficient, I think it has been simplified too far and will break in some cases. I've left more notes in the code in-line so I won't repeat here.

There are also some more minor stylistic changes I've commented on. I didn't comment on every instance so please check for other uses of e.g. auto instead of concrete types or `map.count(idx) -> map[idx]` instead of the more typical `it = map.find(idx) -> it->...`

james-sumihiro committed:
Added support for VK_EXT_nested_command_buffer

Serialisation functions were updated for new structs and enum values. CaptureQueueSubmit was updated to recurse nested command buffer records so that command buffers and their resources are referenced in the capture. The logic for rebasing secondary command buffer events was updated to support nesting. The PartialReplayData struct was updated so all partial submissions can be tracked during active replay.

baldurk commented on renderdoc/driver/vulkan/vk_core.h:
```diff
@@ -653,6 +653,16 @@ class WrappedVulkan : public IFrameCapturer
+    uint32_t beginEvent = 0;
+    rdcarray<CommandBufferNode> executedCommands;
+    VkCommandBufferLevel level = VK_COMMAND_BUFFER_LEVEL_MAX_ENUM;
+    CommandBufferNode *rootNode = nullptr;
```
Minor style note, please use `NULL` rather than nullptr.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp:
```diff
@@ -906,6 +896,50 @@ void WrappedVulkan::InsertActionsAndRefreshIDs(BakedCmdBufferInfo &cmdBufInfo)
+{
+  const rdcarray<VkResourceRecord *> &subcmds = record->bakedCommands->cmdInfo->subcmds;
+
+  for(auto &subcmd : subcmds)
```
This use of `auto &subcmd` I would prefer as `VkResourceRecord *subcmd`.

baldurk commented on renderdoc/driver/vulkan/vk_core.h:
```diff
@@ -735,59 +745,121 @@ class WrappedVulkan : public IFrameCapturer
+  } m_Partial;
 
-  struct Submission
+  rdcflatmap<ResourceId, CommandBufferNode> m_CommandNodes;
```
I think potentially the duplication of CommandBufferNode and arrays-of-arrays here might be a potential future issue, but this is probably fine for now until we see this be a problem. Worth putting a mental pin in it - but use of secondary command buffers is rare and I don't expect that to change, so this will be quite a "pay for what you use" situation.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp:
```diff
@@ -249,6 +249,20 @@ void WrappedVulkan::DoSubmit(VkQueue queue, VkSubmitInfo2 submitInfo)
+                                                CommandBufferNode &rootNode, uint32_t curBaseEvent)
+{
+  parentNode.beginEvent += curBaseEvent;
+  parentNode.rootNode = &rootNode;
```
This pointer is not safe to take and cache because this object is in the `m_Partial.commandTree` array below which may resize and invalidate all cached root node pointers.

I think given the way the other code is structured it would probably be simplest to dynamically allocate the entries in `m_Partial.commandTree` so that they can be cached like this.

baldurk commented on renderdoc/driver/vulkan/vk_core.h:
```diff
@@ -735,59 +745,121 @@ class WrappedVulkan : public IFrameCapturer
+           (m_BakedCmdBufferInfo[cmdId].level == VK_COMMAND_BUFFER_LEVEL_PRIMARY);
+  }
+
+  void SetPartialStack(const CommandBufferNode &targetNode, uint32_t curEvent)
```
As a minor and admittedly quite arbitrary point, could you move these functions `SetPartialStack`, `BuildPartialStackUpToTarget`, and `BuildPartialStack` to the source file? Usually for in-line functions I limit to only accessor-type things - a lookup and a check or something like that. The other functions here are fine to leave in the header.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -1169,50 +1172,33 @@ bool WrappedVulkan::Serialise_vkBeginCommandBuffer(SerialiserType &ser, VkComman
     if(IsActiveReplaying(m_State))
     {
-      const uint32_t length = m_BakedCmdBufferInfo[BakedCommandBuffer].eventCount;
+      rdcarray<CommandBufferNode> submits = m_Partial.submitLookup[BakedCommandBuffer];
```
This should be a `const rdcarray<CommandBufferNode> &submits` to avoid a copy.

baldurk commented on renderdoc/driver/vulkan/vk_core.h:
```diff
@@ -735,59 +745,121 @@ class WrappedVulkan : public IFrameCapturer
     }
+  }
+
+  void BuildPartialStack(const CommandBufferNode &parentNode, uint32_t ev)
```
This function doesn't seem to be used?

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -1780,10 +1763,10 @@ bool WrappedVulkan::Serialise_vkCmdBeginRenderPass(SerialiserType &ser, VkComman
 
         // only if we're partially recording do we update this state
-        if(ShouldUpdateRenderState(m_LastCmdBufferID, true))
+        if(IsCommandBufferPartialPrimary(m_LastCmdBufferID) || m_OutsideCmdBuffer != VK_NULL_HANDLE)
```
I think there is still value in having a small utility function similar to `ShouldUpdateRenderState` to do this logic even if it's minimal, if only to avoid 'leaking' the awareness of `m_OutsideCmdBuffer` into the implementation of command replays like this.

Worst case it's a slightly unnecessary helper function, but in case the command buffer handling is refactored in future it's good to keep the exact replaying details more central.

baldurk reviewed (changes requested):
I think this is looking a lot better, thank you for working on it. I've left some notes as I believe there are still a few issues to resolve but only one is an actual problem the others are stylistic for the most part.

One other general thing I want to say is that I would like you to add some comments to the key structures - `m_CommandNodes`, and `partialStack`, `commandTree`, and `submitLookup` in `m_Partial`. Each one could do with a good comment to explain what it does - I was able to follow the logic through from this PR but once looking in the code and in 6-12 months time it may be less clear.

E.g. it's not obvious that `m_CommandNodes` is only valid/used during loading and is somewhat 'temporary' since beginEvent is overwritten each time. Maybe it should even be cleared after loading?

Similarly `submitLookup` is fairly clear though could use a comment similar to the old lookup that it replaces, but the difference between `commandTree` and `partialStack` isn't necessarily clear from a glance and the name.

baldurk commented:
Manually merged to resolve a trivial conflict and to commit a memory leak fix.
